### DISCUSSION TITLE
How to type parameter as any Prisma model (or perhaps even better, any Prisma model with a field called 'id')

### DISCUSSION THREAD

**Discussion by TeddyBurnsides** [prisma/prisma#9927] [Q&A]
I am attempting to write a function called `makeUniqueModelId()` that generates a unique ID of a specified length for any model with an `id` field. I have a basic framework written below, however I have left the `model` parameter with a type of `any` as I am unsure of how to type it. 

```typescript
const makeUniqueModelId = async (model: any, length: number): Promise<number> => {
  let newId = makeNumber(length);
  // regenerate until we find an ID that isn't in use yet
  while (await model.count({where: {id: newId}}) !== 0) {
    newId = makeNumber(length);
  }
  return newId;
}

const makeNumber = (length: number): number => {
  return Math.floor(Math.random() * Math.pow(10,length));
}
```

How could I type the `model` parameter so that it accepts a Prisma Model? I'd like to use it like so (where `prisma.player` is a model for example): 

```typescript
makeUniqueModelId(prisma.player,6);
```

**Comment by ryands17** [comment]
@theowiersema 👋 

I don't think there's a simple way of typing this. You would need to take the models from the Prisma instance omitting the other things on that instance. So something like:

```ts
type Models = Pick<typeof prisma, 'user' | 'post'>
```

**Reply by TeddyBurnsides** [reply]
@ryands17 That appears to trigger a type error of the following when used like so:

```typescript
const newId = await makeUniqueModelId(prisma.player,6);
```
```typescript
 type ModelsWithId=Pick<typeof prisma, 'player' | 'skillLevel'>
 export const makeUniqueModelId = async (model: ModelsWithId, maxLength: number): Promise<number> => {
    let newId = randomNumber(maxLength);
    // regenerate until we find an ID that isn't in use yet
    while (await model.count({where: {id: newId}}) !== 0) {
      newId = randomNumber(maxLength);
    }
    return newId;
  }
```

> Argument of type 'PlayerDelegate<RejectOnNotFound | RejectPerOperation | undefined>' is not assignable to parameter of type 'ModelsWithId'.
  Type 'PlayerDelegate<RejectOnNotFound | RejectPerOperation | undefined>' is missing the following properties from type 'ModelsWithId': player, skillLevel

**Reply by ryands17** [reply]
Could you try this instead of the above:

```ts
type Models = 'user' | 'post'
```

**Reply by TeddyBurnsides** [reply]
I was able to use

```typescript
type Models = typeof prisma.player
```

Although as soon as I add the other model it throws an error on the `.count` query:

```typescript
 type ModelsWithId=typeof prisma.skillLevel | typeof prisma.player;
 export const makeUniqueModelId = async (model: ModelsWithId, maxLength: number): Promise<number> => {
    let newId = randomNumber(maxLength);
    // regenerate until we find an ID that isn't in use yet
    while (await model.count({where: {id: newId}}) !== 0) {
      newId = randomNumber(maxLength);
    }
    return newId;
  }
```

> This expression is not callable.
  Each member of the union type '(<T extends { where?: SkillLevelWhereInput | undefined; orderBy?: Enumerable<SkillLevelOrderByWithRelationInput> | undefined; ... 4 more ...; select?: true | ... 1 more ... | undefined; }>(args?: Subset<...> | undefined) => PrismaPromise<...>) | (<T extends { ...; }>(args?: Subset<...> | undefined) => PrismaPromise<...' has signatures, but none of those signatures are compatible with each other.

Somehow the `.count` query is different for each model? 

Ultimately it would be cool if I just check if the model has an `id` field but that may not be possible with Prisma.

**Reply by ryands17** [reply]
Yes the return type is different and the args are also typed differently for each model, so it's a bit complicated. Have a look at this thread for a more in-depth discussion: https://github.com/prisma/prisma/issues/5273

**Reply by TeddyBurnsides** [reply]
Thanks @ryands17 that's helpful! That particular discussion really is what I'm getting at (perhaps at a more general level though.)

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
